boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+	modeenv, err := ReadModeenv(dirs.GlobalRootDir)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: unable to read modeenv: %v", err)
+	}
+
+	if modeenv.Base == "" {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: modeenv base boot variable is empty")
+	}
+
+	sn, err := snap.ParsePlaceInfoFromSnapFileName(modeenv.Base)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: modeenv base boot variable is invalid: %v", err)
+	}
+
+	return sn, nil, false, nil
+}
+
+// TODO:UC20: implement this
+func (s20 *bootState20Base) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	return update, nil
+}
+
+// TODO:UC20: implement this with modeenv
+func (s20 *bootState20Base) setNext(nextKernel snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	return false, nil, nil
+}
+
+type bootState20Kernel struct{}
+
+func newBootState20(typ snap.Type) bootState {
+	switch typ {
+	case snap.TypeKernel:
+		return &bootState20Kernel{}
+	case snap.TypeBase:
+		return &bootState20Base{}
+	default:
+		panic("unsupported snap type update")
+	}
+}
+
+func (s20 *bootState20Kernel) revisions() (theSnap, theTrySnap snap.PlaceInfo, trying bool, err error) {
+	var bootSnap, tryBootSnap snap.PlaceInfo
+
+	// the trying snap is the one given by the try-kernel.efi symlink, while
+	// the known-good snap is the one given by the kernel.efi symlink
+	// we can determine if we are trying the try snap by the presence of
+	// 1. a try-kernel.efi symlink
+	// 2. kernel_status == trying
+
+	bl, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot find any bootloader: %v", err)
+	}
+
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if !ok {
+		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
+	}
+
+	// get the kernel for this bootloader
+	bootSnap, err = ebl.Kernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate try kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	if tryKernelExists {
+		tryBootSnap = tryKernel
+	}
+
+	m, err := bl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	trying = (m["kernel_status"] == "trying")
+
+	return bootSnap, tryBootSnap, trying, nil
+}
+
+func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	var u20 *bootStateUpdate20
+	var err error
+	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+	if err != nil {
+		return nil, err
+	}
+
+	// kernel_status goes from "" -> "try" -> "trying" -> ""
```

Comment:
Given the level of details this has, we could probably add that `"trying"` is set by the boot script.